File: backend/app/core/parser/cpp.py
import re
from pathlib import Path
from typing import List

from app.api.models import ImportInfo


class CppParser:
    """Parser for C and C++ files."""

    def get_supported_extensions(self) -> List[str]:
        """Get supported file extensions."""
        return [".c", ".h", ".cpp", ".hpp", ".cc", ".hh", ".cxx"]

    def parse_file(self, file_path: Path) -> List[ImportInfo]:
        """Parse a C/C++ file and extract include statements.

        Args:
            file_path: Path to the file

        Returns:
            List of ImportInfo objects
        """
        imports = []
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()

            # Updated regex to handle:
            # - Spaces between # and include (# include)
            # - No space between include and " or < (#include<stdio.h>)
            patterns = [
                (r'#\s*include\s*"([^"]+)"', "local"),
                (r'#\s*include\s*<([^>]+)>', "system"),
            ]

            for i, line in enumerate(content.split("\n"), 1):
                line = line.strip()
                if line.startswith("//"):
                    continue

                for pattern, import_type in patterns:
                    matches = re.finditer(pattern, line)
                    for match in matches:
                        imports.append(
                            ImportInfo(
                                source_file=str(file_path),
                                imported_module=match.group(1),
                                import_type=import_type,
                                line_number=i,
                            )
                        )
        except Exception as e:
            print(f"CppParser error: {e}")
            pass
        return imports

File: extractor/extract/cpp.go
package extract

import (
	"path/filepath"
	"regexp"
	"strings"
)

var (
	cppLocalIncludeRegex  = regexp.MustCompile(`#\s*include\s*"([^"]+)"`)
	cppSystemIncludeRegex = regexp.MustCompile(`#\s*include\s*<([^>]+)>`)
	cppExtensions         = []string{".c", ".h", ".cpp", ".hpp", ".cc", ".hh", ".cxx"}
)

// IsCppFile checks if a file is a C/C++ file based on extension.
func IsCppFile(path string) bool {
	ext := strings.ToLower(filepath.Ext(path))
	for _, e := range cppExtensions {
		if ext == e {
			return true
		}
	}
	return false
}

// ExtractCpp extracts includes from a C/C++ source file.
func ExtractCpp(content string, filePath string) []Import {
	var imports []Import
	lines := strings.Split(content, "\n")

	patterns := map[*regexp.Regexp]string{
		cppLocalIncludeRegex:  "local",
		cppSystemIncludeRegex: "system",
	}

	for i, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "//") {
			continue
		}

		for re, importType := range patterns {
			matches := re.FindAllStringSubmatch(line, -1)
			for _, match := range matches {
				if len(match) > 1 {
					imports = append(imports, Import{
						SourceFile:     filePath,
						ImportedModule: match[1],
						ImportType:     importType,
						LineNumber:     i + 1,
					})
				}
			}
		}
	}
	return imports
}